quantum/encoder.c

```diff
@@ -81,12 +81,35 @@ void encoder_init(void) {
     }
 #endif
 
+#ifdef ENC_PUP
     for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
         setPinInputHigh(encoders_pad_a[i]);
         setPinInputHigh(encoders_pad_b[i]);
 
         encoder_state[i] = (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
     }
+#elif ENC_PDOWN
+    for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
+        setPinInputLow(encoders_pad_a[i]);
+        setPinInputLow(encoders_pad_b[i]);
+
+        encoder_state[i] = (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
+    }
+#elif ENC_FLOAT
+    for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
+        setPinInput(encoders_pad_a[i]);
+        setPinInput(encoders_pad_b[i]);
+
+        encoder_state[i] = (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
+    }
+#else
+    for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
+        setPinInputHigh(encoders_pad_a[i]);
+        setPinInputHigh(encoders_pad_b[i]);
+
+        encoder_state[i] = (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
+    }
+#endif
```

Comment:
For example, the following makes your code simple and easy to maintain.

Add the following code to `quantum/config_common.h`. 
```c
/* encoder pin mode */
#define ENC_PUP   0
#define ENC_PDOWN 1
#define ENC_FLOAT 2
```

The user configures the following in `config.h`: 
```c
#define ENCODER_PIN_MODE ENC_PUP /* or ENC_PDOWN, ENC_FLOAT */
```

```suggestion
#ifdef ENCODER_PIN_MODE
#  if ENCODER_PIN_MODE == ENC_PUP
#    define setEncPinInput(x) setPinInputHigh(x)
#  elif ENCODER_PIN_MODE == ENC_PDOWN
#    define setEncPinInput(x) setPinInputLow(x)
#  elif ENCODER_PIN_MODE == ENC_FLOAT
#    define setEncPinInput(x) setPinInput(x)
#  endif
#else
#  define setEncPinInput(x) setPinInputHigh(x)
#endif

    for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
        setEncPinInput(encoders_pad_a[i]);
        setEncPinInput(encoders_pad_b[i]);

        encoder_state[i] = (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
    }
```

Reply:
Could even include: 
```c
#if defined(ENCODER_PIN_PDOWN)
#    ifdef __AVR__
#        Error Pull Down is not supported on AVR
#    endif 
#    define setEncPinInput(x) setPinInputLow(x)
#elif defined(ENCODER_PIN_FLOAT)
#    define setEncPinInput(x) setPinInput(x)
#elif defined(ENCODER_PIN_OUTPUT
#    define setEncPinInput(x) setPinOutput(x)
#else
#    define setEncPinInput(x) setPinInputHigh(x)
#endif
```

Not sure if output would even work, though.